### PR TITLE
Qthreads:  Add spack smoke test

### PR DESCRIPTION
**Pull request by janciesko**

### CODE REVIEW

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 1]
```diff
@@ -85,3 +85,58 @@ def configure_args(self):
                     % self.spec.variants['scheduler'].value)
 
         return args
+
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py:91]: FWIW. From a static analysis standpoint, there should be only 1 line between the `return` and the properties.  This issue and others are reported in the flake8 results.

**Comment by tldahlgren** [comment]
FYI.  I'm checking out and running the test tests.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 2]
```diff
@@ -85,3 +85,58 @@ def configure_args(self):
...
+        tests = self.test_list
+        relative_test_dir = self.test_base_path
+
+        tests_to_cpy = []
+        header = 'test/argparsing.h'
+        for test in tests:
+            test_path = join_path(relative_test_dir, test + '.c')
+            tests_to_cpy.append(test_path)
+        
+        self.cache_extra_test_sources(tests_to_cpy)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: You could combine calls to `self.cache_extra_test_sources` to provide a single list with all of the files you want to copy to reduce the copy overhead.  For example, you could change `tests_to_cpy` to `files_to_cpy` and initialize it to a list consisting of `header` before your `for` loop.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 3]
```diff
@@ -85,3 +85,58 @@ def configure_args(self):
...
+        self.cache_extra_test_sources(header) 
+
+    def build_tests(self):
+        """Build and run the added smoke (install) test."""
+        tests = self.test_list
+        relative_test_dir = self.test_base_path
+
+        for test in tests:
+            options = [
+               '-I{0}'.format(self.prefix.include),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: ~~FWIW. I had to add `'-std=c11'` as an option to get this to compile for gcc@8.1.0 on a linux box and actually run the tests BUT this is *not* the correct solution.~~

As I mentioned in email, I get more focused error messages in the test output with the default scheduler *and* the all of the tests pass for me when the library is built with the `distrib` scheduler when the proper compiler is used.  (Thanks for pointing me to it!)

**Comment by tldahlgren** [comment]
This is *really* close!

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 4]
```diff
@@ -85,3 +85,58 @@ def configure_args(self):
                     % self.spec.variants['scheduler'].value)
 
         return args
+
+
+
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]:
```suggestion
    test_requires_compiler = True
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/qthreads/package.py]: You'll need this property set to `True` to get the right compiler for the build.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 5]
```diff
@@ -85,3 +85,58 @@ def configure_args(self):
...
+    test_base_path = 'test/basics/'
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: For readability, the three properties -- `test_requires_compiler` *and* the two properties here -- should be moved to under the `url` property at the top of the package.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Please fix the flake8 failures and add the `test_requires_compiler = True` property.

**Comment by janciesko** [comment]
Comments addressed. Thanks @tldahlgren.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 6]
```diff
@@ -83,5 +87,52 @@ def configure_args(self):
...
+                      '-I{0}'.format(self.prefix.include),
+                      '-I{0}'.format(self.install_test_root + '/test'),
+                      join_path(self.install_test_root, relative_test_dir,
+                               test + '.c'),
+                      '-o',
+                      test,
+                      '-L{0}'.format(self.prefix.lib),
+                      '-lqthread',
+                      '{0}{1}'.format(self.compiler.cc_rpath_arg,
+                                     self.prefix.lib ]
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: You need a closing paren (')') before the closing square bracket here.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 7]
```diff
@@ -83,5 +87,51 @@ def configure_args(self):
...
+        self.cache_extra_test_sources(files_to_cpy)
+
+    def build_tests(self):
+        """Build and run the added smoke (install) test."""
+        tests = self.test_list
+        relative_test_dir = self.test_base_path
+
+        for test in tests:
+            options = [
+                      '-I{0}'.format(self.prefix.include),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: Python flake8 is picky.  Lines 113 to 121 should be indented 4 spaces relative to line 112, with two exceptions.

The first letter of `test` in line 116 should appear under the `s` in line 115.  A similar alignment is needed on line 122.

File: var/spack/repos/builtin/packages/qthreads/package.py [hunk 8]
```diff
@@ -83,5 +87,51 @@ def configure_args(self):
...
+                      '-I{0}'.format(self.install_test_root + '/test'),
+                      join_path(self.install_test_root, relative_test_dir,
+                               test + '.c'),
+                      '-o',
+                      test,
+                      '-L{0}'.format(self.prefix.lib),
+                      '-lqthread',
+                      '{0}{1}'.format(self.compiler.cc_rpath_arg,
+                                     self.prefix.lib)]
+            reason = 'test:{0}: Checking ability to link to the library.'.format(test)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/qthreads/package.py]: You can add a slash ('\\') after the apostrophe in `library.'` so Python knows to continue this statement on the next line to pick up `.format(test)`, which needs to be indented 4 spaces relative to `reason` in line 123.

**Review by tldahlgren — APPROVED** [review]
I'm able to successfully build and run the smoke tests (which pass when I build with `scheduler=distrib`).

This can be merged once the flake8 checks pass.

**Comment by janciesko** [comment]
Flake8 passes. Commits squashed.

**Comment by tldahlgren** [comment]
> Flake8 passes. Commits squashed.

Great!  Thanks.

This PR can be merged once all the required checks pass.

**Comment by tldahlgren** [comment]
Thank you for your contribution!

**Pull request merged**